### PR TITLE
Load the developer certificate from a default path provided a configuration key has been specified for it

### PR DESCRIPTION
**Pull request by javiercn**
Makes kestrel search for a certificate on a default path based on the presence of a password in config for the default certificate and no explicit path configured for it.

This pattern supports containers scenarios where the path changes between the app being run in a container vs the app being run on the host.

### CODE REVIEW

File: src/Kestrel.Core/Internal/KestrelServerOptionsSetup.cs [hunk 1]
```diff
@@ -32,7 +36,36 @@ private void UseDefaultDeveloperCertificate(KestrelServerOptions options)
...
+                if (configReader.Certificates.TryGetValue("Default", out var certificateConfig) &&
+                    certificateConfig.Path == null && certificateConfig.Password != null)
+                {
+                    var hostingEnvironment = options.ApplicationServices.GetRequiredService<IHostingEnvironment>();
+                    var appName = hostingEnvironment.ApplicationName;
+                    var appData = Environment.GetEnvironmentVariable("APPDATA");
+                    var home = Environment.GetEnvironmentVariable("APPDATA");
+                    var basePath = appData != null ? $"{appData}/ASP.NET/https" : null;
+                    basePath = basePath ?? (home != null ? $"{home}/.aspnet/https" : null);
+                    var certificatePath = basePath != null ? $"{basePath}/{appName}.pfx" : null;
```
Comment (Tratcher) [inline src/Kestrel.Core/Internal/KestrelServerOptionsSetup.cs]: appName.pfx?

File: samples/SampleApp/appsettings.Production.json [hunk 2]
```diff
@@ -9,6 +9,11 @@
           "Password": "testPassword"
         }
       }
+    },
+    "Certificates": {
+      "Default": {
```
Comment (Tratcher) [inline samples/SampleApp/appsettings.Production.json]: Use a different node with it's own defined semantics. E.g. dockercert or appdatacert.

File: test/Kestrel.Tests/Kestrel.Tests.csproj [hunk 3]
```diff
@@ -14,6 +14,7 @@
   <ItemGroup>
     <ProjectReference Include="..\..\src\Kestrel\Kestrel.csproj" />
     <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
+    <!--<PackageReference Include="Microsoft.AspNetCore.Certificates.Generation.Sources" Version="$(MicrosoftAspNetCoreCertificatesGenerationSourcesPackageVersion)" />-->
```
Comment (Tratcher) [inline test/Kestrel.Tests/Kestrel.Tests.csproj]: ?

File: test/Kestrel.Tests/KestrelConfigurationBuilderTests.cs [hunk 4]
```diff
@@ -209,5 +212,117 @@ public void ConfigureEndpointDefaultCanEnableHttps()
...
+                var path = GetCertificatePath();
+                Directory.CreateDirectory(Path.GetDirectoryName(path));
+                File.WriteAllBytes(path, bytes);
+
+                var ran1 = false;
+                var config = new ConfigurationBuilder().AddInMemoryCollection(new[]
+                {
+                new KeyValuePair<string, string>("Endpoints:End1:Url", "https://*:5001"),
+                new KeyValuePair<string, string>("Certificates:Development:Password", "1234"),
+            }).Build();
```
Comment (Tratcher) [inline test/Kestrel.Tests/KestrelConfigurationBuilderTests.cs]: formatting

File: test/Kestrel.Tests/KestrelConfigurationBuilderTests.cs [hunk 5]
```diff
@@ -209,5 +212,117 @@ public void ConfigureEndpointDefaultCanEnableHttps()
...
+                var certificate = new X509Certificate2(TestResources.GetCertPath("aspnetdevcert.pfx"), "aspnetdevcert", X509KeyStorageFlags.Exportable);
+                var bytes = certificate.Export(X509ContentType.Pkcs12, "1234");
+                var path = GetCertificatePath();
+                Directory.CreateDirectory(Path.GetDirectoryName(path));
+                File.WriteAllBytes(path, bytes);
+
+                var config = new ConfigurationBuilder().AddInMemoryCollection(new[]
+                {
+                new KeyValuePair<string, string>("Certificates:Development:Password", "12341234"),
+            }).Build();
```
Comment (Tratcher) [inline test/Kestrel.Tests/KestrelConfigurationBuilderTests.cs]: formatting

File: src/Kestrel.Core/KestrelConfigurationLoader.cs [hunk 6]
```diff
@@ -259,6 +263,76 @@ private void LoadDefaultCert(ConfigurationReader configReader)
...
+                }
+            }
+
+            return false;
+        }
+
+        private bool TryGetCertificatePath(out string path)
+        {
+            var hostingEnvironment = Options.ApplicationServices.GetRequiredService<IHostingEnvironment>();
+            var appName = hostingEnvironment.ApplicationName;
```
Comment (Tratcher) [inline src/Kestrel.Core/KestrelConfigurationLoader.cs]: The use of app name as the cert name still doesn't make sense. This is one more variable that can make this fail. Why can't you use a single instance named aspnetdevcert? Everything else is hard coded.

File: test/Kestrel.Tests/Kestrel.Tests.csproj (hunk 3 above)
Comment (javiercn) [reply test/Kestrel.Tests/Kestrel.Tests.csproj]: This is just leftover, I'll clean it up

File: src/Kestrel.Core/KestrelConfigurationLoader.cs (hunk 6 above)
Comment (javiercn) [reply src/Kestrel.Core/KestrelConfigurationLoader.cs]: Discussed offline

File: src/Kestrel.Core/KestrelConfigurationLoader.cs [hunk 7]
```diff
@@ -259,6 +263,76 @@ private void LoadDefaultCert(ConfigurationReader configReader)
...
+
+            // This will go away when we implement
+            // https://github.com/aspnet/Hosting/issues/1294
```
Comment (muratg) [inline src/Kestrel.Core/KestrelConfigurationLoader.cs]: After this gets checked in, could you put a note on that bug to remind to remove this?

Comment (javiercn) [reply src/Kestrel.Core/KestrelConfigurationLoader.cs]: Yep

**Pull request merged**